### PR TITLE
Separate module for base distributions

### PR DESCRIPTION
**Pull request by jmarshrossney**
Created a module called distributions.py which currently contains just one class `NormalDist`.

This class does the following
- Handles generation of standard normal variates through the `__call__` method, which is used during both training and sampling.
- Contains method to compute the log density which is used in the loss function during training only.

The `NormDist` object is passed to `RealNVP` under the name `generator`, which I'm not entirely happy with.

It may seem redundant, but I'm finding that this system is easily extensible, since it means we don't have to have base distributions hard-coded into models.

Subsequent PR's will add another class, to distributions.py which deals with uniform distributions on N-spheres, containing methods for e.g. calculating the Jacobian for the spherical parameterisation.

### CODE REVIEW

File: anvil/distributions.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        self.n_batch = n_batch
+        self.lattice_volume = lattice_volume
+        self.field_dimension = field_dimension
+
+        # Number of components per field configuration =
+        # size of output Tensor at dimension 1
+        self.size_out = self.lattice_volume * self.field_dimension
+
+        # Pre-calculate normalisation for log density
+        self._log_normalisation = self.log_normalisation()
```
Comment (wilsonmr) [inline anvil/distributions.py]: probably more idiomatic to have the function start with underscore because its a "private" method and the attribute would be public

File: anvil/distributions.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    def log_normalisation(self) -> float:
+        """logarithm of the normalisation for the density function."""
+        print(self.size_out)
+        return log(sqrt(pow(2 * pi, self.size_out)))
+
+    def log_density(self, sample: torch.Tensor) -> torch.Tensor:
+        """Return log probability density of a sample generated from
+        the __call__ method above with default arguments!
+
+        The size of the sample (number of configurations) should be exactly
```
Comment (wilsonmr) [inline anvil/distributions.py]: I don't understand, log normalisation is independent of batch size and so it should be, right?

Comment (wilsonmr) [reply anvil/distributions.py]: in fact I think n_batch is completely unneccessary in this class?

File: anvil/distributions.py (hunk 1 above)
Comment (jmarshrossney) [reply anvil/distributions.py]: ah yes, this is true.

**Comment by wilsonmr** [comment]
seems like a good idea, I like it. Just look at my comments then we can merge

File: anvil/distributions.py (hunk 2 above)
Comment (jmarshrossney) [reply anvil/distributions.py]: Also very true. Outdated comment.

Comment (jmarshrossney) [reply anvil/distributions.py]: cheers for picking up on that

**Comment by jmarshrossney** [comment]
Apologies for many small commits. Assuming we can "squash" them all during the merge?

**Comment by wilsonmr** [comment]
urm yeah you can do an interactive rebase (git rebase -i) you have to select a commit and then a text editor will open and you can select to either pick or squash commits and then you get a chance to edit the commit messages as well

**Comment by jmarshrossney** [comment]
what about just using the "squash and merge" feature on the drop down menu next to "Merge pull request"?

**Comment by wilsonmr** [comment]
oooh fancy, yeah that also works

**Comment by jmarshrossney** [comment]
are you happy for me to do this or would you prefer to maintain exclusive merging privileges for the master branch?

File: anvil/config.py [hunk 3]
```diff
@@ -158,10 +173,10 @@ def parse_sample_interval(self, interval: (int, type(None))):
         log.warning(f"Using user specified sample_interval: {interval}")
         return interval
 
-    def parse_bootstrap_n_samples(self, n_samples: int):
+    def parse_n_boot(self, n_boot: int):
```
Comment (wilsonmr) [inline anvil/config.py:176]: you rename this here but then I see that `observables.two_point_function` is still taking the old input probably needs to be consistent everywhere

**Comment by wilsonmr** [comment]
So I don't mind who merges but I've added a repo so that merging is blocked unless there is approval, just to avoid any upsets. I think anyone can approve the PR so it's not really a restriction.. Also as a collaborator I think you probably can overwrite this anyway

As soon as you look at the bootstrap thing then I'm happy for this to be merged and you can handle the squash merge

File: anvil/config.py (hunk 3 above)
Comment (jmarshrossney) [reply anvil/config.py:176]: Ah yes.
We currently have `n_boot` in the example runcard, though maybe that was by accident.
I'll change everything to `n_boot` unless you'd rather keep `bootstrap_n_sample` - it's probably clear enough what it means.

Comment (wilsonmr) [reply anvil/config.py:176]: not bothered - change to n_boot

**Review by wilsonmr — APPROVED** [review]

**Comment by wilsonmr** [comment]
cool I'll leave the merging to you
